### ISSUE TITLE
Clarify behavior of setResourceTimingBufferSize

### ISSUE THREAD

**Issue by JosephPecoraro** [w3c/resource-timing#96]
[Spec](https://w3c.github.io/resource-timing/#dom-performance-setresourcetimingbuffersize) for setResourceTimingBufferSize is very simple:

> The setResourceTimingBufferSize method runs the following steps:
>
> 1. Set resource timing buffer size limit to the maxSize parameter. If the maxSize parameter is less than resource timing buffer current size, no PerformanceResourceTiming objects are to be removed from the performance entry buffer.
> 2. Set resource timing buffer full flag to false.

However, most implementations do not appear to matches this.

Do we want to change the spec or should I just file bugs?

---

Test case:

```js
function loadRandomResource() {
    return fetch(window.location.href + "?" + Math.random());
}

// Get the browser into a consistent state.
performance.clearResourceTimings();
performance.setResourceTimingBufferSize(100);

// Gather up 3 Resource Entries to kick off the rest of test behavior.
let observer = new PerformanceObserver(function(list) {
    let allResourceEntries = performance.getEntriesByType("resource");
    if (allResourceEntries.length !== 3)
        return;
    observer.disconnect();
    test();
});
observer.observe({entryTypes: ["resource"]});
for (let i = 0; i < 3; ++i)
    loadRandomResource();

// Test behavior of setting buffer limit to the current size (3).
function test() {
    performance.onresourcetimingbufferfull = function() {
        console.log("Event Fired", performance.getEntriesByType("resource").length);
    };
    console.log("before setLimit(3)");
    performance.setResourceTimingBufferSize(3);
    console.log("after setLimit(3)");
    loadRandomResource().then(function() {
        console.log("after loading 4th resource");
    });
}
```

* Expected:

    before setLimit(3)
    after setLimit(3)
    Event Fired - 3
    after loading 4th resource

* Chrome dispatches the buffer full event inside setResourceTimingBufferSize if the new limit is less than or equal to the current size. [(source)](https://chromium.googlesource.com/chromium/blink/+/master/Source/core/timing/PerformanceBase.cpp#167)

    before setLimit(3)
    Event Fired - 3
    after setLimit(3)

* Firefox doesn't maintain separate limit and "buffer full flag". So when the 4th resource loads it is already over the limit so it doesn't dispatch an event. [(source)](https://hg.mozilla.org/mozilla-central/file/tip/dom/performance/Performance.cpp)

    before setLimit(3)
    after setLimit(3)
    after loading 4th resource

* Edge - ?

**Comment by yoavweiss** [comment]
Converted @JosephPecoraro's test case into a WPT: https://github.com/web-platform-tests/wpt/pull/11551

**Comment by yoavweiss** [comment]
Also, IIUC, we don't really need further clarifications in the spec, just to file bugs against non-compliant implementations, right?

**Comment by wanderview** [comment]
> Also, IIUC, we don't really need further clarifications in the spec, just to file bugs against non-compliant implementations, right?

Well, the spec doesn't say to fire an event if the limit is lowered passed the current length, right?  Are you saying browsers should or should not fire an event in that case?

FWIW, there is a service worker test that depends on firing the event in this case:

https://github.com/web-platform-tests/wpt/blob/4321fff7672396a92873a3e4b36cf2305c56369a/service-workers/service-worker/resources/performance-timeline-worker.js#L49

I'm trying to determine what to do about that.  I guess I will probably make that test not depend on this behavior as its unclear to me what the final spec decision will be.  Also, it doesn't seem to be the primary purpose of that test.

**Comment by wanderview** [comment]
Oh, we don't dispatch bufferfull event on workers at all right now because we are waiting for #141 to be resolved.

**Comment by yoavweiss** [comment]
@wanderview - I think the spec is clear that such an event should not be fired (although it doesn't explicitly says that, just explicitly states that entries should not be removed).

We could add a clarification if you think one is needed, or we could revisit that decision if you think there's a compelling use case for firing an event in that case. (IMO, developers can simply check the buffer size before setting it to a value that's below the current size, so an event is not extremely useful)

**Comment by yoavweiss** [comment]
@npm1 - re-reading the `setResourceTimingBufferSize` [processing](https://w3c.github.io/resource-timing/#dom-performance-setresourcetimingbuffersize), isn't the current model dropping the entry which fires the event? If so, it might be better to fire the event earlier, before entries have to be dropped? (and in which case, to actually *do* fire them when the value is set lower than current size, just for consistency)

**Comment by wanderview** [comment]
@yoavweiss I'm fine with the current spec (no strong opinion either way), but the service worker WPT test probably needs to be fixed as part of this.

**Comment by npm1** [comment]
Yea I agree it might be better to fire the onresourcetimingbufferfull event right away when it becomes full due to changing the size of the buffer.

**Comment by toddreifsteck** [comment]
Ideally, we should not be speccing or implementing re-entrant behavior. It appears that Chrome currently implements in this way.

Can this problem be solved by altering the spec and all UAs to work without allowing this while still solving the problem that this event exists for?

**Comment by yoavweiss** [comment]
Closed by #168
